ec2/src/main/scala/awscala/ec2/InstanceWithKeyPair.scala

```diff
@@ -11,22 +14,21 @@ case class InstanceWithKeyPair(
 
   import com.decodified.scalassh._
 
-  def ssh[T](f: SshClient => SSH.Result[T]) = SSH[T](publicDnsName, provider(keyPairFile))(f)
+  def ssh[T](f: SshClient => SSH.Result[T]): Try[T] = SSH[T](publicDnsName, provider(keyPairFile))(f)
 
-  private[this] def provider(keyPairFile: File): HostConfigProvider = new FromStringsHostConfigProvider {
-    def rawLines(host: String): com.decodified.scalassh.Validated[(String, TraversableOnce[String])] =
-      if (keyPairFile.exists()) {
-        Right("dummy_source" -> (
-          Seq(
-            "login-type = keyfile",
-            s"username = $user",
-            s"keyfile = ${keyPairFile.getAbsolutePath}",
-            s"command-timeout = $connectionTimeout",
-            "fingerprint = any" //TODO: ask if user will trust any host key provided by the server. Currently it's always YES.
-          )))
-      } else {
-        Left(s"KeyFile ${keyPairFile.getAbsolutePath} does not exist")
-      }
+  private[this] def provider(keyPairFile: File): HostConfigProvider = new HostConfigProvider {
+    override def apply(v1: String): Try[HostConfig] =
```

Comment:
The implementation here has been completely changed. I haven't checked it out yet but have you made sure if it works with real EC2 resources? If it works for sure, it seems to look good to me.